src/dataframe/dataframe.jl

```diff
@@ -1806,3 +1806,94 @@ function _replace_columns!(df::DataFrame, newdf::DataFrame)
     copy!(index(df).lookup, index(newdf).lookup)
     return df
 end
+
+expandgrid(; kwargs...) = isempty(kwargs) ? DataFrame() : expandgrid(kwargs...)
+
+expandgrid(pairs::Pair{<:AbstractString, <:Any}...) =
+    expandgrid((Symbol(k) => v for (k,v) in pairs)...)
+
+"""
+    expandgrid(pairs::Pair...)
+    expandgrid(kwargs...)
+
+Create a `DataFrame`` from all combinations of the passed arguments.
+
+It is allowed to pass  a list of `Pair`s as positional
+arguments, or a list of keyword arguments. Each `Pair`` is considered
+to represent a column name to column values to expand mapping.
+Column name must be a `Symbol` or string. All passed column names must be unique.
+
+Column value can be a vector which is consumed as is or an object of any other
+type (except `AbstractArray`). In the latter case the passed value is treated
+as having length one for expansion. As a particular rule values stored in a `Ref`
+or a `0`-dimensional `AbstractArray` are unwrapped and treated as having length one.
+
+`DataFrame` can store only columns that use 1-based indexing. Attempting
+to store a vector using non-standard indexing after `repeat` is called on it
+will raise an error.
+
+# Examples
+
+```jldoctest
+julia> expandgrid(a=1:2, b='a':'c')
+6×2 DataFrame
+ Row │ a      b
+     │ Int64  Char
+─────┼─────────────
+   1 │     1  a
+   2 │     2  a
+   3 │     1  b
+   4 │     2  b
+   5 │     1  c
+   6 │     2  c
+
+julia> expandgrid("a" => 1:2, "b" => 'a':'c', "c" => "const")
+6×3 DataFrame
+ Row │ a      b     c
+     │ Int64  Char  String
+─────┼─────────────────────
+   1 │     1  a     const
+   2 │     2  a     const
+   3 │     1  b     const
+   4 │     2  b     const
+   5 │     1  c     const
+   6 │     2  c     const
+```
+"""
+function expandgrid(pairs::Pair{Symbol, <:Any}...)
+    colnames = first.(pairs)
+    if !allunique(colnames)
+        throw(ArgumentError("All column names passed to expandgrid must be unique"))
+    end
+    colvalues = map(pairs) do p
+        v = last(p)
+        if v isa AbstractVector
+            return v
+        elseif v isa Union{AbstractArray{<:Any, 0}, Ref}
+            x = v[]
+            return fill!(Tables.allocatecolumn(typeof(x), 1), x)
```

Comment:
```suggestion
            return fill!(Tables.allocatecolumn(typeof(x), 1), v[])
```
Should we use `eltype(v)` instead of `typeof(x)`? Just a thought. We already use this pattern elsewhere, right?

Reply:
I believe that the code that we have now is correct (i.e. I would not apply your suggestion).

The point is that if you have:
```
julia> x = Any[1]
1-element Vector{Any}:
 1

julia> y = @view x[1]
0-dimensional view(::Vector{Any}, 1) with eltype Any:
1

julia> eltype(y)
Any

julia> z = y[]
1

julia> typeof(z)
Int64
```
then you cannot identify the type properly with `eltype` of the container. Similarly:
```
julia> q = Ref{Any}(1)
Base.RefValue{Any}(1)

julia> eltype(q)
Any

julia> p = q[]
1

julia> typeof(p)
Int64
```

Reply:
Yeah it's correct, I was just wondering whether it would make sense to preserve the eltype of the container when it's different from that of the eltype, as there may be a reason for this difference. For example, `Ref{Union{Int, Missing}}(0)` would create a column which allows for missing values. But I'm not sure that's a good reason to do it.

Reply:
We currently do not do it in other places when doing pseudo broadcasting:
```
julia> DataFrame(a=Ref{Union{Int,Missing}}(1))
1×1 DataFrame
 Row │ a
     │ Int64
─────┼───────
   1 │     1

julia> combine(groupby(DataFrame(a=1:3), :a), :a => x -> Ref{Union{Int, Missing}}(x[1]))
3×2 DataFrame
 Row │ a      a_function
     │ Int64  Int64
─────┼───────────────────
   1 │     1           1
   2 │     2           2
   3 │     3           3
```
so I now implemented it to be consistent with other places.

Reply:
OK then at least we're consistent.